File: cortex/webgraph/graph_builder.py
from __future__ import annotations

from collections import Counter

from cortex.webgraph.config import WebGraphConfig
from cortex.webgraph.contracts import (
    EpisodicRecord,
    SemanticRecord,
    WebGraphEdge,
    WebGraphNode,
    WebGraphSnapshot,
    WebGraphStats,
)
from cortex.webgraph.relation_builder import RelationBuilder


class GraphBuilder:
    """Assemble the hybrid Cortex graph from semantic and episodic records."""

    def __init__(self, config: WebGraphConfig) -> None:
        self.config = config
        self.relation_builder = RelationBuilder(config)

    def build_snapshot(
        self,
        *,
        fingerprint: str,
        mode: str,
        semantic_records: list[SemanticRecord],
        episodic_records: list[EpisodicRecord],
    ) -> WebGraphSnapshot:
        if mode == "semantic":
            episodic_records = []
        elif mode == "episodic":
            semantic_records = []

        edges = self.relation_builder.build_edges(semantic_records, episodic_records)
        nodes = self._build_nodes(semantic_records, episodic_records, edges)
        visible_node_ids = {node.id for node in nodes}
        filtered_edges = [
            edge
            for edge in edges
            if edge.source in visible_node_ids and edge.target in visible_node_ids
        ]
        stats = WebGraphStats(
            node_count=len(nodes),
            edge_count=len(filtered_edges),
            mode=mode,  # type: ignore[arg-type]
            truncated=False,
        )
        return WebGraphSnapshot(
            fingerprint=fingerprint,
            mode=mode,  # type: ignore[arg-type]
            stats=stats,
            nodes=nodes,
            edges=filtered_edges,
        )

    @staticmethod
    def _build_nodes(
        semantic_records: list[SemanticRecord],
        episodic_records: list[EpisodicRecord],
        edges: list[WebGraphEdge],
    ) -> list[WebGraphNode]:
        degree_counter: Counter[str] = Counter()
        for edge in edges:
            degree_counter[edge.source] += 1
            degree_counter[edge.target] += 1

        nodes: list[WebGraphNode] = []
        for s_record in semantic_records:
            nodes.append(
                WebGraphNode(
                    id=s_record.node_id,
                    node_type=s_record.node_type,
                    source="semantic",
                    label=s_record.title,
                    summary=s_record.summary,
                    rel_path=s_record.rel_path,
                    tags=list(s_record.tags),
                    degree=degree_counter.get(s_record.node_id, 0),
                    metadata={"abs_path": s_record.abs_path},
                )
            )

        for e_record in episodic_records:
            nodes.append(
                WebGraphNode(
                    id=e_record.node_id,
                    node_type=e_record.node_type,
                    source="episodic",
                    label=e_record.label,
                    summary=e_record.summary,
                    memory_id=e_record.memory_id,
                    tags=list(e_record.tags),
                    files=list(e_record.files),
                    timestamp=e_record.timestamp,
                    degree=degree_counter.get(e_record.node_id, 0),
                    metadata=dict(e_record.metadata),
                )
            )

        return sorted(nodes, key=lambda node: (node.source, node.node_type, node.label.lower()))

File: rust/crates/cortex-webgraph-server/src/graph_builder.rs
//! Porteo de `cortex/webgraph/graph_builder.py`.

#![forbid(unsafe_code)]

use std::collections::HashMap;

use crate::config::WebGraphConfig;
use crate::contracts::{
    EpisodicRecord, SemanticRecord, WebGraphEdge, WebGraphNode, WebGraphSnapshot, WebGraphStats,
};
use crate::relation_builder::RelationBuilder;

pub struct GraphBuilder {
    pub config: WebGraphConfig,
}

impl GraphBuilder {
    pub fn new(config: WebGraphConfig) -> Self {
        Self { config }
    }

    pub fn build_snapshot(
        &self,
        fingerprint: &str,
        mode: &str,
        generated_at: &str,
        mut semantic_records: Vec<SemanticRecord>,
        mut episodic_records: Vec<EpisodicRecord>,
    ) -> WebGraphSnapshot {
        if mode == "semantic" {
            episodic_records.clear();
        } else if mode == "episodic" {
            semantic_records.clear();
        }
        let relation = RelationBuilder::new(self.config.clone());
        let edges = relation.build_edges(&semantic_records, &episodic_records);
        let nodes = Self::build_nodes(&semantic_records, &episodic_records, &edges);
        let visible: std::collections::HashSet<&str> =
            nodes.iter().map(|n| n.id.as_str()).collect();
        let filtered_edges: Vec<WebGraphEdge> = edges
            .into_iter()
            .filter(|e| visible.contains(e.source.as_str()) && visible.contains(e.target.as_str()))
            .collect();
        let stats = WebGraphStats {
            node_count: nodes.len() as i64,
            edge_count: filtered_edges.len() as i64,
            mode: mode.to_string(),
            truncated: false,
        };
        WebGraphSnapshot {
            version: "2.0".into(),
            fingerprint: fingerprint.to_string(),
            generated_at: generated_at.to_string(),
            mode: mode.to_string(),
            stats,
            capabilities: Default::default(),
            nodes,
            edges: filtered_edges,
        }
    }

    fn build_nodes(
        semantic_records: &[SemanticRecord],
        episodic_records: &[EpisodicRecord],
        edges: &[WebGraphEdge],
    ) -> Vec<WebGraphNode> {
        let mut degree_counter: HashMap<&str, i64> = HashMap::new();
        for edge in edges {
            *degree_counter.entry(edge.source.as_str()).or_insert(0) += 1;
            *degree_counter.entry(edge.target.as_str()).or_insert(0) += 1;
        }

        let mut nodes: Vec<WebGraphNode> = Vec::new();
        for s in semantic_records {
            // metadata + abs_path encima (campo legacy).
            let mut node_meta = s.metadata.clone();
            node_meta.insert("abs_path".into(), serde_json::json!(s.abs_path));
            nodes.push(WebGraphNode {
                id: s.node_id.clone(),
                node_type: s.node_type.clone(),
                source: "semantic".into(),
                label: s.title.clone(),
                summary: s.summary.clone(),
                rel_path: Some(s.rel_path.clone()),
                memory_id: None,
                tags: s.tags.clone(),
                files: vec![],
                timestamp: None,
                degree: degree_counter.get(s.node_id.as_str()).copied().unwrap_or(0),
                metadata: node_meta,
            });
        }
        for e in episodic_records {
            nodes.push(WebGraphNode {
                id: e.node_id.clone(),
                node_type: e.node_type.clone(),
                source: "episodic".into(),
                label: e.label.clone(),
                summary: e.summary.clone(),
                rel_path: None,
                memory_id: Some(e.memory_id.clone()),
                tags: e.tags.clone(),
                files: e.files.clone(),
                timestamp: e.timestamp.clone(),
                degree: degree_counter.get(e.node_id.as_str()).copied().unwrap_or(0),
                metadata: e.metadata.clone(),
            });
        }
        nodes.sort_by(|a, b| {
            (
                a.source.as_str(),
                a.node_type.as_str(),
                a.label.to_lowercase(),
            )
                .cmp(&(
                    b.source.as_str(),
                    b.node_type.as_str(),
                    b.label.to_lowercase(),
                ))
        });
        nodes
    }
}
